torch/onnx/_internal/fx/export.py

```diff
@@ -17,22 +19,96 @@ def _onnx_not_available(name):
             "ONNX is not available. Please install ONNX to use this feature."
         )
 
-    onnx.__getattr__ = _onnx_not_available
+    onnx.__getattr__ = _onnx_not_available  # type: ignore[assignment]
 
 import torch
 import torch._C
 import torch._decomp
 import torch._dynamo
 import torch._ops
+import torch.fx
 from torch._subclasses import fake_tensor
 from torch.fx.experimental import proxy_tensor
 from torch.fx.passes import fake_tensor_prop
 from torch.nn.utils import stateless
 from torch.onnx._globals import GLOBALS as ONNX_GLOBALS
-from torch.onnx._internal import jit_utils, registration
+from torch.onnx._internal import _beartype, jit_utils, registration
 from torch.utils import _pytree
 
 
+class ModuleExpansionTracer(torch.fx._symbolic_trace.Tracer):
+    """Tracer to create ONNX-exporting friendly FX graph.
+
+    This tracer traces models into operators. That is,
+    the traced graph mostly contains call_function nodes and
+    has no call_module nodes. The call_module nodes
+    are problematic to the use of make_fx(...) in ONNX
+    exporter.
+    """
+
+    @_beartype.beartype
+    def is_leaf_module(self, m: torch.nn.Module, module_qualified_name: str) -> bool:
+        return False
+
+    @_beartype.beartype
+    def to_bool(self, obj: "torch.fx.Proxy") -> bool:
+        # This is a hack to tracing through if-else Python blocks.
+        # It may generate incorrect ONNX graphs if the if-else block
+        return False
+
+
+@_beartype.beartype
+def module_expansion_symbolic_trace(
+    root: Union[torch.nn.Module, Callable[..., Any]],
+    concrete_args: Optional[Dict[str, Any]] = None,
+) -> "torch.fx.GraphModule":
+    _TORCH_METHODS_TO_PATCH = ["arange", "tensor", "finfo", "full", "empty"]
+
+    def gen_constructor_wrapper(target):
+        @functools.wraps(target)
+        def wrapper(*args, **kwargs):
+            proxy = None
+
+            def check_has_proxy(v):
+                if isinstance(v, torch.fx.Proxy):
+                    nonlocal proxy
+                    proxy = v
+
+            torch.fx.node.map_aggregate(args, check_has_proxy)
+            torch.fx.node.map_aggregate(kwargs, check_has_proxy)
+
+            if proxy is not None:
+                return proxy.tracer.create_proxy("call_function", target, args, kwargs)
+            else:
+                return target(*args, **kwargs)
+
+        return wrapper, target
+
+    patched_torch_methods = {
+        target: gen_constructor_wrapper(getattr(torch, target))
+        for target in _TORCH_METHODS_TO_PATCH
+    }
+    orig_fns = set()
+
+    for name, (wrapper, orig) in patched_torch_methods.items():
+        setattr(torch, name, wrapper)
+        orig_fns.add(orig)
+
+    try:
```

Comment:
What are the exceptions we expect?

Reply:
No exception expected. It just ensures the patches will be reverted in
```py
    finally:
        # Revert the patches for symbolic tracing.
        for name, (_, wrapped) in patched_torch_methods.items():
            # wrapped is the original version of `torch.name`.
            setattr(torch, name, wrapped)
```